### PR TITLE
 [RFC] crimson/net: potential optimizations and evaluation

### PR DESCRIPTION
**Pull request by cyx1231st**
Got 34.6% better IOPS in average.

Optimizations ideas (PoC implementation with v1):
* [A] gather message buffers when send:
  - Gather buffers from pending messages and send them together;
  - Minimize the attempts to check keepalive/keepalive_ack when sending;
  - Reduce seastar tasks in the write path;
* [B] batch message reads:
  - Keep decoding outside the fast read path;
  - Batch message decodings;
  - Reduce seastar tasks in the read path;
* [C] increase out buffer to 65536
  - Improve batching in the write path;
* [D] set `seastar::net::posix_data_source_impl` with `buf_size = 65536` (NOT included)
  - Improve batching in the read path;

[E] Batching heavily uses `seastar::promise<>`, so applied optimizations https://github.com/rzarzynski/seastar/commit/682bd9053b50ff9c09e270b534b612d53618ce0c and https://github.com/rzarzynski/seastar/commit/682bd9053b50ff9c09e270b534b612d53618ce0c from @rzarzynski 

Test command:
```
$perf_crimson_msgr --round=2097152 --cbs=4096 --sbs=4096 -c 3
perf settings:
  client[>> v1:0.0.0.0:9010/0](bs=4096, rounds=2097152, jobs=1, depth=512)
  server[v1:0.0.0.0:9010/0](bs=4096, core=0)
```

Test result (RelWithDebInfo build):
```
rounds [master] [A~D] [A~D]+E
1      10.97s   8.23s 8.09s
2      10.92s   8.20s 8.08s
3      10.82s   8.26s 8.10s
4      10.87s   8.20s 8.09s
5      10.93s   8.23s 8.15s
```

### CODE REVIEW

File: src/crimson/net/ProtocolV1.cc [hunk 1]
```diff
@@ -595,6 +595,45 @@ void ProtocolV1::start_accept(SocketFRef&& sock,
 
 // open state
 
+seastar::future<> ProtocolV1::write_messages(std::queue<MessageRef>& msgs)
+{
+  bufferlist bl;
+  size_t len = msgs.size();
+  while (!msgs.empty()) {
+    auto msg = msgs.front();
```
Comment (rzarzynski) [inline src/crimson/net/ProtocolV1.cc]: Maybe there is no need to bump up the counter of `MessageRef`. We could `::pop()` after processing current message.

File: src/crimson/net/ProtocolV1.cc [hunk 2]
```diff
@@ -595,6 +595,45 @@ void ProtocolV1::start_accept(SocketFRef&& sock,
 
 // open state
 
+seastar::future<> ProtocolV1::write_messages(std::queue<MessageRef>& msgs)
+{
+  bufferlist bl;
+  size_t len = msgs.size();
```
Comment (rzarzynski) [inline src/crimson/net/ProtocolV1.cc]: nit: unused?

File: src/test/crimson/perf_async_msgr.cc [hunk 3]
```diff
@@ -31,9 +31,16 @@ struct Server {
...
-    {}
+    unsigned msg_len = 0;
+    bufferlist msg_data;
+
+    ServerDispatcher(CephContext* cct, unsigned msg_len)
+      : Dispatcher(cct), msg_len(msg_len)
+    {
+      bufferptr ptr(msg_len);
+      memset(ptr.c_str(), 0, msg_len);
+      msg_data.append(ptr);
```
Comment (rzarzynski) [inline src/test/crimson/perf_async_msgr.cc]: See `ceph::bufferlist::append_zero`.

File: src/crimson/net/ProtocolV1.cc [hunk 4]
```diff
@@ -595,6 +595,45 @@ void ProtocolV1::start_accept(SocketFRef&& sock,
 
 // open state
 
+seastar::future<> ProtocolV1::write_messages(std::queue<MessageRef>& msgs)
+{
+  bufferlist bl;
```
Comment (rzarzynski) [inline src/crimson/net/ProtocolV1.cc]: You might want to `::reserve()` required space for the C-string taking `::append()` calls:
```cpp
  bl.reserve(msgs.size() * (sizeof(CEPH_MSGR_TAG_MSG) + sizeof(header) + sizeof(footer)));
```

This should prevent from allocating 4k at first call and wasting memory when the queue has few messages.

File: src/crimson/net/Protocol.cc [hunk 5]
```diff
@@ -114,12 +114,8 @@ seastar::future<> Protocol::write_worker()
         return seastar::now();
       }).then([this] {
         if (!conn.out_q.empty()){
-          MessageRef msg = conn.out_q.front();
-          return write_message(msg)
-          .then([this, msg] {
-            if (msg == conn.out_q.front()) {
-              conn.out_q.pop();
-            }
+          return write_messages(conn.out_q)
```
Comment (rzarzynski) [inline src/crimson/net/Protocol.cc]: :+1:

File: src/crimson/net/ProtocolV1.cc (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/net/ProtocolV1.cc]: done.

File: src/crimson/net/ProtocolV1.cc (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/net/ProtocolV1.cc]: removed

File: src/test/crimson/perf_async_msgr.cc (hunk 3 above)
Comment (cyx1231st) [reply src/test/crimson/perf_async_msgr.cc]: done.

File: src/crimson/net/ProtocolV1.cc (hunk 4 above)
Comment (cyx1231st) [reply src/crimson/net/ProtocolV1.cc]: Sorry not familiar with bufferlist interface, can I still use `bl.append(...)` to fill the C-strings, after calling `bl.reserve(...)`?

Comment (rzarzynski) [reply src/crimson/net/ProtocolV1.cc]: Yup, it's OK.

**Comment by cyx1231st** [comment]
Close this in favor of https://github.com/ceph/ceph/pull/27836 https://github.com/ceph/ceph/pull/27788 and https://github.com/ceph/seastar/pull/4

**Pull request closed without merging**